ddht/v5_1/alexandria/client.py

```diff
@@ -0,0 +1,130 @@
+
+from eth_typing import NodeID
+
+from ddht.endpoint import Endpoint
+from ddht.exceptions import DecodingError
+from ddht.request_tracker import RequestTracker
+from ddht.v5_1.abc import NetworkAPI
+from ddht.v5_1.alexandria.abc import AlexandriaClientAPI
+from ddht.v5_1.alexandria.constants import ALEXANDRIA_PROTOCOL_ID
+from ddht.v5_1.alexandria.messages import (
+    AlexandriaMessage,
+    PingMessage,
+    PongMessage,
+    TAlexandriaMessage,
+    decode_message,
+)
+from ddht.v5_1.alexandria.payloads import PingPayload, PongPayload
+
+
+class AlexandriaClient(AlexandriaClientAPI):
+    protocol_id = ALEXANDRIA_PROTOCOL_ID
+
+    _active_request_ids: Set[bytes]
+
+    def __init__(self, network: NetworkAPI) -> None:
+        self.network = network
+        self.request_tracker = RequestTracker()
+
+        self._active_request_ids = set()
+
+    #
+    # Request/Response message sending primatives
+    #
+    async def _send_request(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        message: AlexandriaMessage[Any],
+        *,
+        request_id: Optional[bytes] = None,
+    ) -> bytes:
+        data_payload = message.to_wire_bytes()
+        request_id = await self.network.client.send_talk_request(
+            node_id,
+            endpoint,
+            protocol=ALEXANDRIA_PROTOCOL_ID,
+            payload=data_payload,
+            request_id=request_id,
+        )
+        return request_id
+
+    async def _send_response(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        message: AlexandriaMessage[Any],
+        *,
+        request_id: bytes,
+    ) -> None:
+        data_payload = message.to_wire_bytes()
+        await self.network.client.send_talk_response(
+            node_id, endpoint, payload=data_payload, request_id=request_id,
+        )
+
+    async def _request(
+        self,
+        node_id: NodeID,
+        endpoint: Optional[Endpoint],
+        request: AlexandriaMessage[Any],
+        response_class: Type[TAlexandriaMessage],
+    ) -> TAlexandriaMessage:
+        # The `TalkResponseMessage` only includes a `request_id` but not the
+        # actual `protocol_id`. In order to implement a subscription API for
+        # the Alexandria sub-protocol messages, we need to be able to identify
+        # which responses are part of the Alexandria sub-protocol.  The
+        # `request_id` is the mechanism we use to accomplish this.  We manually
+        # supply the `request_id` for requests so that we can use a separate
+        # `RequestTrackerAPI`.  Then, we can subscribe to **all**
+        # `TalkResponseMessage` messages, and only feed them through the
+        # subscription when the `request_id` is known by the tracker.
+        request_id = self.network.client.request_tracker.get_free_request_id(node_id)
+        request_data = request.to_wire_bytes()
+        with self.request_tracker.reserve_request_id(node_id, request_id):
```

Comment:
It seems really strange that on 82 we use the client's request tracker, but here we use our own request tracker. Doesn't this mean that other other modules which talk to that client might try to use this request_id? We're only reserving it locally so they won't know not to use this request id.

Reply:
If I'm reading this correctly `self.network.client.dispatcher.subscribe_request` already does dispatching based on the `request_id`, so this class shouldn't need to be worried about maintaining its own `request_id` namespace. Maybe I'm misreading something?

Reply:
The comment above was supposed to explain why this is happening but it appears I didn't do a very adequate job.

We need a subscription API for alexandria messages in order to be able to cleanly implement things like "send a pong anytime someone sends us a ping".  In order to do this we need to be able to monitor all TALKREQUEST and TALKRESPONSE messages for this protocol.  The TALKREQUEST message is easy since it contains a `protocol_id`.  For the TALKRESPONSE however, the only way for us to know that an incoming message belongs to this protocol is to match it against an in-flight request (which we implicitely know is part of the alexandria protocol).

So, this use of a separate `RequestTrackerAPI` is the way we can accomplish this.  We still acquire the `request_id` from the core request tracker (on line 82), so we get the guarantee that we are indeed using a unique `request_id`.  Then, we track that `request_id` in this Alexandria-specific tracker, so that we can monitor incoming TALKRESPONSE messages, check if their `request_id` is tracked, and if so: attempt to decode it as an alexandria message and feed it into any active subscriptions.

Reply:
I also recognize that this is probably more confusing since the functionality that is actually using this isn't introduced until #118 (the next PR in line)

Reply:
Ok, I updated the comment, I'm curious to get any feedback on how I might make that clearer.